File: 007. Reverse Integer/solution-logn.js
/**
 * @param {number} x
 * @return {number}
 */

// Time: O(logN)
var reverse = function(x) {
    var max = Math.pow(2, 31) - 1;
    var min = Math.pow(-2, 31);
    
    if (x > max || x < min) return 0;
    
    var ret = 0;
    var isNeg = x < 0 ? true : false;
    x = Math.abs(x);
    
    while (x > 0) {
        var digit = x % 10;
        if (!isNeg && 10 * ret > max - digit) {
            return 0;
        } else if (isNeg && -10 * ret < min + digit) {
            return 0;
        } else {
            ret = 10 * ret + digit;
        }
        
        x = ~~(x / 10);
    }
    
    return isNeg ? -ret : ret;
};

File: Leet-007. Reverse Integer/solution-logn.java
class Solution {
    public int reverse(int x) {
        int max = Integer.MAX_VALUE;
        int min = Integer.MIN_VALUE;
        
        boolean isNeg = x < 0 ? true : false;
        x = Math.abs(x);
        int res = 0;
        
        while (x > 0) {
            int digit = x % 10;
            
            if (max / 10 < res || max / 10 == res && max % 10 < digit) {
                return 0;
            }
            
            res = 10 * res + digit;
            x /= 10;
        }
        
        return isNeg ? -res : res;
    }
}